docs/design/variadics.md

```diff
@@ -0,0 +1,1018 @@
+   binding map {}.
+3. (By binding introduction) `T` equals `i32` given a binding map {(`T`,
+   `i32`)}.
+4. (By parameterized type equality) `Vector(T)` equals `Vector(i32)` given a
+   binding map {(`T`, `i32`)}.
+5. (By equality implies convertibility) `Vector(T)` is convertible to
+   `Vector(i32)` given a binding map {(`T`, `i32`)}.
+
+Notice that as a byproduct, this proof gives us a binding map with the deduced
+values of the parameters.
+
+Conceptually, the typechecker constructs this proof by searching backwards from
+desired conclusion without considering binding maps, and then running the
+resulting proof forward to compute the final binding map.
+
+### Deduction algorithm
+
+A full pattern is in _normal form_ if it contains no pack literals, and every
+arity coercion is fully expanded. Note that all user-written full patterns are
+in normal form. Note also that by construction, this means that the type of the
+body of every pack expansion has a single scalar component. The _canonical form_
+of a full pattern is the unique normal form (if any) that is "maximally merged",
+meaning that every tuple pattern and tuple literal has the smallest number of
+segments.
+
+> **TODO:** Specify algorithm for converting a full pattern to canonical form,
+> or establishing that there is no such form. See next section for a start.
+
+If a function with type `F` is called with argument type `A`, we typecheck the
+call by converting `F` to canonical form, and then checking whether `A` is
+convertible to the parameter type by applying the deduction rules in the
+previous section. If that succeeds, we apply the resulting binding map to the
+function return type to obtain the type of the call expression.
+
+> **TODO:** Specify the algorithm more precisely. In particular, discuss how to
+> rewrite `A` as needed to make the shapes line up, but don't rewrite `F` after
+> canonicalization.
+
+Typechecking for pattern match operations other than function calls is defined
+in terms of typechecking a function call: We check a scrutinee type `S` against
+a pattern `P` by checking `__F(S,)` against a hypothetical function signature
+`fn __F(P,)->();`.
+
+> **Future work:** Extend this approach to support merging the argument list as
+> well as the parameter list.
+
+#### Canonicalization algorithm
+
+The canonical form can be found by starting with a normal form, and
+incrementally merging an adjacent singular parameter type into the variadic
+parameter type.
+
+For example, consider the following function:
+
+```carbon
+fn F[First:! type, Second:! type, ... each Next:! type]
+    (first: Vector(First), second: Vector(Second),
+     ... each next: Vector(each Next)) -> (First, Second, ... each Next);
+```
+
+First, we desugar the implicit arity:
+
+```carbon
+fn F[__N:! Arity, First:! type, Second:! type, ... each Next:! ⟪type; __N⟫]
+    (first: Vector(First), second: Vector(Second),
+     ... each next: Vector(each Next)) -> (First, Second, ... each Next);
```

Comment:
(continued from before)
In this example, there are two pack expansion patterns (`... each Next` and `... each next`), but only one implicit arity.